File: 백준/Silver/9012. 괄호/괄호.py
import sys 
def valid(l):
  stack = []
  for ch in l:
    if ch == '(':
      stack.append(ch)
    else:
      if stack:
        stack.pop()
      else:
        if ch == ')':
          return "NO" 
  if stack:
    return "NO"
  else:
    return "YES"

t = int(sys.stdin.readline())
for _ in range(t):
  line = sys.stdin.readline().strip()
  print(valid(line))
  
      
  

File: 백준/Silver/9012. 괄호/괄호.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
let [n, ...arr] = fs.readFileSync(filePath).toString().trim().split("\n");

const N = +n;
const ps = arr.map((pair) => pair.split(""));

const answer = [];

for (let i = 0; i < N; i++) {
  const stack = [];
  for (let [idx, p] of ps[i].entries()) {
    stack.push(p);

    if (p === ")") {
      if (idx === 0) {
        break;
      }
      if (stack.at(-2) === "(") {
        stack.pop();
        stack.pop();
      }
    }
  }
  if (stack.length === 0) answer.push("YES");
  else answer.push("NO");
}

console.log(answer.join("\n"));
